sunbeam-python/sunbeam/plugins/validation/plugin.py

```diff
@@ -0,0 +1,304 @@
+# Copyright (c) 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from datetime import datetime
+from typing import Optional
+
+import click
+from croniter import croniter
+from packaging.version import Version
+from rich.console import Console
+
+from sunbeam.clusterd.client import Client
+from sunbeam.clusterd.service import ClusterServiceUnavailableException
+from sunbeam.commands.openstack import OPENSTACK_MODEL
+from sunbeam.jobs.juju import JujuHelper, run_sync
+from sunbeam.plugins.interface.v1.openstack import (
+    OpenStackControlPlanePlugin,
+    TerraformPlanLocation,
+)
+
+LOG = logging.getLogger(__name__)
+console = Console()
+
+PLUGIN_VERSION = "0.0.1"
+MINIMAL_PERIOD = 15 * 60
+TEMPEST_CHANNEL = "latest/edge"
+VALIDATION_PLUGIN_DEPLOY_TIMEOUT = 60 * 60  # tempest can take some time to initialized
+
+
+def validate_schedule(schedule: str) -> tuple[bool, Optional[str]]:
+    """Validate if the schedule config option is validate or not."""
+    # Empty schedule is fine; it means it's disabled in this context.
+    if not schedule:
+        return True, "Schedule is an empty string, disabling the periodic check."
+
+    # croniter supports second repeats, but vixie cron does not.
+    if len(schedule.split()) == 6:
+        return False, "This cron does not support seconds in schedule (6 fields)."
+
+    # constant base time for consistency
+    base = datetime(2004, 3, 5)
+
+    try:
+        cron = croniter(schedule, base, max_years_between_matches=1)
+    except ValueError as e:
+        msg = str(e)
+        # croniter supports second repeats, but vixie cron does not,
+        # so update the error message here to suit.
+        if "Exactly 5 or 6 columns" in msg:
+            msg = "Exactly 5 columns must be specified for iterator expression."
+        return False, msg
+
+    # This is a rather naive method for enforcing this,
+    # and it may be possible to craft an expression
+    # that results in some consecutive runs within 15 minutes,
+    # however this is fine, as there is process locking for tempest,
+    # and this is more of a sanity check than a security requirement.
+    t1 = cron.get_next()
+    t2 = cron.get_next()
+    if t2 - t1 <= MINIMAL_PERIOD:
+        return False, "Schedule repeats faster than 15 minutes."
```

Comment:
Need to give better error message here.
```suggestion
        return False, "Cannot schedule periodic check to run faster than every 15 minutes."
```

Reply:
DONE